File: app/src/main/java/co/garmax/materialflashlight/features/SettingsRepository.java
package co.garmax.materialflashlight.features;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import javax.inject.Inject;

import co.garmax.materialflashlight.features.modes.ModeBase;
import co.garmax.materialflashlight.features.modules.ModuleBase;

public class SettingsRepository {

    private static final String KEEP_SCREEN_ON = "keep_screen_on";
    private static final String MODE = "mode";
    private static final String MODULE = "module";
    private static final String AUTO_TURN_ON = "auto_turn_on";

    private SharedPreferences sharedPreferences;

    @Inject
    public SettingsRepository(Context context) {
        sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context);
    }

    public boolean isKeepScreenOn() {
        return sharedPreferences.getBoolean(KEEP_SCREEN_ON, false);
    }

    public void setKeepScreenOn(boolean isKeepScreenOn) {
        sharedPreferences.edit().putBoolean(KEEP_SCREEN_ON, isKeepScreenOn).apply();
    }

    public boolean isAutoTurnOn() {
        return sharedPreferences.getBoolean(AUTO_TURN_ON, false);
    }

    public void setAutoTurnOn(boolean isAutoTurnOn) {
        sharedPreferences.edit().putBoolean(AUTO_TURN_ON, isAutoTurnOn).apply();
    }

    public ModeBase.Mode getMode() {
        String mode = sharedPreferences.getString(MODE, ModeBase.Mode.MODE_TORCH.toString());

        return ModeBase.Mode.valueOf(mode);
    }

    public void setMode(ModeBase.Mode mode) {
        sharedPreferences.edit().putString(MODE, mode.name()).apply();
    }

    public ModuleBase.Module getModule() {
        String module = sharedPreferences.getString(MODULE,
                ModuleBase.Module.MODULE_CAMERA_FLASHLIGHT.toString());

        return ModuleBase.Module.valueOf(module);
    }

    public void setModule(ModuleBase.Module module) {
        sharedPreferences.edit().putString(MODULE, module.name()).apply();
    }
}

File: app/src/main/java/co/garmax/materialflashlight/features/SettingsRepository.kt
package co.garmax.materialflashlight.features

import android.content.Context
import co.garmax.materialflashlight.features.modes.ModeBase
import co.garmax.materialflashlight.features.modules.ModuleBase

class SettingsRepository(context: Context) {

    private val sharedPreferences = context.getSharedPreferences("settings", Context.MODE_PRIVATE)

    var isKeepScreenOn: Boolean
        get() = sharedPreferences.getBoolean(KEEP_SCREEN_ON, false)
        set(isKeepScreenOn) {
            sharedPreferences.edit().putBoolean(KEEP_SCREEN_ON, isKeepScreenOn).apply()
        }

    var isAutoTurnedOn: Boolean
        get() = sharedPreferences.getBoolean(AUTO_TURN_ON, false)
        set(isAutoTurnOn) {
            sharedPreferences.edit().putBoolean(AUTO_TURN_ON, isAutoTurnOn).apply()
        }

    var mode: ModeBase.Mode
        get() {
            val mode = sharedPreferences.getString(MODE, null)
            return mode?.let { ModeBase.Mode.valueOf(it) } ?: ModeBase.Mode.MODE_TORCH
        }
        set(mode) {
            sharedPreferences.edit().putString(MODE, mode.name).apply()
        }

    var module: ModuleBase.Module
        get() {
            val module = sharedPreferences.getString(MODULE, null)
            return module?.let { ModuleBase.Module.valueOf(it) }
                ?: ModuleBase.Module.MODULE_CAMERA_FLASHLIGHT
        }
        set(module) {
            sharedPreferences.edit().putString(MODULE, module.name).apply()
        }

    companion object {
        private const val KEEP_SCREEN_ON = "keep_screen_on"
        private const val MODE = "mode_name"
        private const val MODULE = "module_name"
        private const val AUTO_TURN_ON = "auto_turn_on"
    }
}